utils/perf.c

```diff
@@ -133,13 +133,28 @@ int setup_perf_record(struct uftrace_perf_writer *perf, int nr_cpu, int pid,
  */
 void finish_perf_record(struct uftrace_perf_writer *perf)
 {
-	int cpu;
+	int cpu, f_no;
+	char proclnk[PATH_MAX], filename[PATH_MAX];
+	ssize_t r;
+	struct stat stbuf;
 
 	for (cpu = 0; cpu < perf->nr_event; cpu++) {
 		close(perf->event_fd[cpu]);
 		munmap(perf->page[cpu], PERF_MMAP_SIZE);
-		if (perf->fp[cpu])
+		if (perf->fp[cpu]) {
+			f_no = fileno(perf->fp[cpu]);
+			sprintf(proclnk, "/proc/self/fd/%d", f_no);
+			r = readlink(proclnk, filename, PATH_MAX);
+			if (r < 0) {
+				pr_warn("failed to find perf record: %m\n");
+				continue;
+			}
+			filename[r] = '\0';
```

Comment:
I think you'd better adding `output_dir` to `uftrace_perf_writer` and making filenames using cpu index.

Reply:
Did you mean below?
```cpp
struct uftrace_perf_writer {
	int			*event_fd;
	char			*output_dir;
	void			**page;
...
```

Reply:
yep

Reply:
I think that will be better. Thank you for the comment.

Reply:
Then you can simple stat the file without calling `fileno() + readlink()`:

```c
for (cpu = 0; cpu < perf->nr_events; cpu++) {
   ...
   snprintf(filename, sizeof(filename), "%s/perf-cpu%d.dat", perf->output_dir, cpu);
   if (!stat(filename, &stbuf) && stbuf.st_size == 0)
      remove(filename);
}
```